extensions/odl/capabilities/src/main/java/org/mqnaas/extensions/odl/capabilities/impl/ODLRootResourceProvider.java

```diff
@@ -0,0 +1,233 @@
+	}
+
+	@Override
+	public void deactivate() {
+		log.info("Removing ODLRootResourceProvider from resource " + resource.getId());
+		resources.clear();
+		log.info("Removed ODLRootResourceProvider from resource " + resource.getId());
+	}
+
+	@Override
+	public List<IRootResource> getRootResources() {
+		log.info("Getting all RootResources");
+		return new ArrayList<IRootResource>(resources);
+	}
+
+	@Override
+	public List<IRootResource> getRootResources(Type type, String model, String version) throws ResourceNotFoundException {
+		log.info("Getting al RootResources with filter [type=" + type + ",model=" + model + ",version=" + version + "]");
+
+		List<IRootResource> filteredResources = new ArrayList<IRootResource>();
+
+		for (IRootResource resource : getRootResources()) {
+
+			Specification specification = resource.getDescriptor().getSpecification();
+
+			boolean matches = true;
+			matches &= type != null ? specification.getType().equals(type) : true;
+			matches &= model != null ? specification.getModel().equals(model) : true;
+			matches &= version != null ? specification.getVersion().equals(version) : true;
+
+			if (matches)
+				filteredResources.add(resource);
+
+		}
+
+		log.debug("Found " + filteredResources.size() + " resources matching filter [type=" + type + ",model=" + model + ",version=" + version + "]");
+
+		return filteredResources;
+	}
+
+	@Override
+	public IRootResource getRootResource(String id) throws ResourceNotFoundException {
+
+		if (StringUtils.isEmpty(id))
+			throw new NullPointerException("Id of the resource to be found can't be null.");
+
+		for (IRootResource resource : resources) {
+			if (StringUtils.equals(id, resource.getId()))
+				return resource;
+		}
+
+		throw new ResourceNotFoundException("No resource found with id: " + id);
+	}
+
+	@Override
+	public void setRootResources(Collection<IRootResource> rootResources) {
+		throw new UnsupportedOperationException("Service not implemented for ODL networks.");
+	}
+
+	private void initializeResources() throws InstantiationException, IllegalAccessException, CapabilityNotFoundException,
+			ApplicationNotFoundException {
+		log.info("Initializing ODL resources for ODL network " + resource.getId());
+
+		resources = new CopyOnWriteArrayList<IRootResource>();
+
+		Nodes nodes = odlClient.getNodes(DEFAULT_CONNECTOR_NAME);
+
+		for (NodeProperties nodeProperties : nodes.getNodeProperties()) {
+			// get information from the node
+			String nodeId = nodeProperties.getNode().getNodeID();
+			Type nodeType = parseNodeType(nodeProperties.getNode().getNodeType());
+
+			// initialize new RootResource representing this node.
+			Collection<Endpoint> odlResourceEndpoints = new ArrayList<Endpoint>(resource.getDescriptor().getEndpoints());
+			IRootResource odlResource = new RootResource(RootResourceDescriptor.create(new Specification(nodeType), odlResourceEndpoints));
+			rmListener.resourceAdded(odlResource, this, IRootResourceProvider.class);
```

Comment:
I'm not sure whether resources should be added to IRootResourceProvider or to IRootResourceAdministration instead.
Are you sure resources of other types are added to IRootResourceProvider?
(Just a question :))